### PR TITLE
austin.intermediate.summer2.mcmillan.christina

### PR DESCRIPTION
**Pull request by nina2cool**

### CODE REVIEW

File: homework/01LessonOne.js [hunk 1]
```diff
@@ -106,7 +108,7 @@ var notNotTrue;
 // What is the difference between the "and" and the "or" operators? Explain why you would use each of them.
 //
 // Your Answer Goes Here:
-//
+//ANSWER:  the && tests whether the value on the left is true, and if it is, it will evaluate if the value on the right is true.  || tests whether either item is true.
```
Comment (tjefferson08) [inline homework/01LessonOne.js:111]: Great!

File: homework/01LessonOne.js [hunk 2]
```diff
@@ -203,220 +205,220 @@ var stringFourStrictNotEqualsNumberFour;
...
 //
 // Your Answer Goes Here:
-//
+// ANSWER:  == is for weak equality (checks for loose equality); === is for strict equality.
 //
 // What’s the difference between != and !==?
 //
 // Your Answer Goes Here:
-//
+//  ANSWER:  != is not equal; !== is not equal value or not equal type
```
Comment (tjefferson08) [inline homework/01LessonOne.js:213]: Almost! It's a bit simpler than that
!= is the inverse comparison of == (aka not equal)
!== is the inverse comparision of === (you should think of this as not-strict-equal)

File: apps/01PigLatin.js [hunk 3]
```diff
@@ -7,11 +7,80 @@ prompt.start();
...
+// Make sure you run this app with node in the console before answering this question.
+// Also, read Lesson 1 in the textbook.
+// http://intermediate.austincodingacademy.com/_book/lesson_one.html
+//
+// What does prompt.get() do for us?
+//
+// Your Answer:
+// ANSWER:  it allows the user of the app to enter any word they want so it is not hardcoded into the code
 function getPrompt() {
     prompt.get(['word'], function (error, result) {
```
Comment (paulrenenichols) [inline apps/01PigLatin.js:85]: Great job answering the questions, @nina2cool!

toast

File: homework/02LessonTwo.js [hunk 4]
```diff
@@ -54,22 +74,23 @@ function lessThanOrEqualTo(parameter1, parameter2) {
...
 //
 // Your Answer Goes Here:
+//a function is a block of code designed to execute a particular task.  you define a function by starting with the keyword "function" followed by a name, followed by parentheses ()
 //
 //
 // What is a return value?
 //
 // Your Answer Goes Here:
-//
+//the return value is the result of the function
```
Comment (paulrenenichols) [inline homework/02LessonTwo.js:83]: Nice!

File: homework/02LessonTwo.js [hunk 5]
```diff
@@ -54,22 +74,23 @@ function lessThanOrEqualTo(parameter1, parameter2) {
...
 //
 // Your Answer Goes Here:
-//
+//you define a function by starting with the keyword "function" followed by a name, followed by parentheses ()
 //
 // What is a parameter? What is an argument? Is there a difference between the two?
 //
 // Your Answer Goes Here:
-//
+//A parameter is like the placeholder for the values used in the function.  an argument is the actual numbers used when calling the function.  they are similar, but parameter is more generic.
```
Comment (paulrenenichols) [inline homework/02LessonTwo.js:93]: Nice answer!

File: apps/02RockPaperScissors.js [hunk 6]
```diff
@@ -7,7 +7,40 @@ prompt.start();
 
 function rockPaperScissors(hand1, hand2) {
 
-    // Write code here
+    //first, going to make sure hand1 and hand2 are both lowercase
+    hand1 = hand1.toLowerCase();
+    hand2 = hand2.toLowerCase();
```
Comment (paulrenenichols) [inline apps/02RockPaperScissors.js:12]: Great job using **toLowerCase()**!

File: apps/02RockPaperScissors.js [hunk 7]
```diff
@@ -7,7 +7,40 @@ prompt.start();
...
+        if(hand2 === 'paper') {
+          return 'Hand one wins!';
+        }
+        else {
+          return 'Hand two wins!';
+        }
+    }
+    else {
+      return "try again, your inputs don't match";
+    }
```
Comment (paulrenenichols) [inline apps/02RockPaperScissors.js:31]: Nice indenting! Great use of **if** and **else**!

File: homework/02LessonTwo.js [hunk 8]
```diff
@@ -26,6 +26,12 @@ var alwaysTrue = function () {
 // argument1 is equal to argument2
 
 function equals(argument1, argument2) {
+  if(argument1 === argument2) {
```
Comment (tjefferson08) [inline homework/02LessonTwo.js:29]: Protip: return the boolean expression directly! WHAAAAAA!?

File: homework/02LessonTwo.js [hunk 9]
```diff
@@ -34,17 +40,31 @@ function equals(argument1, argument2) {
 // parameter1 is less than or equal to parameter2
 
 function lessThanOrEqualTo(parameter1, parameter2) {
+  if(parameter1 <= parameter2) {
```
Comment (tjefferson08) [inline homework/02LessonTwo.js:43]: Ditto here.

File: homework/02LessonTwo.js [hunk 10]
```diff
@@ -54,22 +74,23 @@ function lessThanOrEqualTo(parameter1, parameter2) {
 // What is a function? How do you define a function in Javascript?
 //
 // Your Answer Goes Here:
+//a function is a block of code designed to execute a particular task.  you define a function by starting with the keyword "function" followed by a name, followed by parentheses ()
```
Comment (tjefferson08) [inline homework/02LessonTwo.js:77]: Boom

File: homework/02LessonTwo.js [hunk 11]
```diff
@@ -118,6 +169,20 @@ function whichSpecies(character) {
...
+  if(number % 4 === 0) {
+    return 'divisible by 4';
+  } else if (number % 2 === 0) {
+    return 'divisible by 2';
+  } else if (number % 3 === 0) {
+    return 'divisible by 3';
+  } else if (number % 5 === 0) {
+    return 'divisible by 5';
+  } else {
+    return 'not divisible by 2, 3, 4, or 5.  try a different number';
```
Comment (tjefferson08) [inline homework/02LessonTwo.js:182]: Nice choice here!

File: homework/03LessonThree.js [hunk 12]
```diff
@@ -20,18 +20,18 @@ var firstName = 'susan';
...
-var valueOfLastName;
+var valueOfLastName = 'plaid';
 
 // Problem 3:
 // Fix the following function so that it returns the first argument
 function returnTheFirstArgument(firstArgument) {
-  var firstArgument = 'bob';
   return firstArgument;
 }
+// THIS DOESN'T LOOK WRONG...
```
Comment (tjefferson08) [inline homework/03LessonThree.js:34]: Works for me!

File: homework/03LessonThree.js [hunk 13]
```diff
@@ -41,7 +41,7 @@ function returnTheFirstArgument(firstArgument) {
 // What is scope?
 //
 // Your Answer Goes Here:
-//
+// scope refers to where the item is accessible from - either global (accessible anywhere) or local (only accessible within a function).
```
Comment (tjefferson08) [inline homework/03LessonThree.js:44]: Good, keep it simple 👍

File: homework/03LessonThree.js [hunk 14]
```diff
@@ -79,7 +79,7 @@ var problemSevenAnswer;
 // What is the ternary operator?  How does it differ from normal conditional statements?
 //
 // Your Answer Goes Here:
-//
+//  the ternary operator is a quick way to return results based on a value being true or false without using if statements.  It's syntax is shorter than an if statement.
```
Comment (tjefferson08) [inline homework/03LessonThree.js:82]: quick is right! readability is key

File: apps/02RockPaperScissors.js [hunk 15]
```diff
@@ -7,10 +7,72 @@ prompt.start();
 
 function rockPaperScissors(hand1, hand2) {
 
-    // Write code here
+    //first, going to make sure hand1 and hand2 are both lowercase
+    hand1 = hand1.toLowerCase();
+    hand2 = hand2.toLowerCase();
+
+    if(hand1 !== 'rock' && hand1 !== 'paper' && hand1 !== 'scissors') {
```
Comment (tjefferson08) [inline apps/02RockPaperScissors.js:14]: Nice validations

File: apps/02RockPaperScissors.js [hunk 16]
```diff
@@ -7,10 +7,72 @@ prompt.start();
...
+      return "invalid entry for hand1 - must be either paper, rock, or scissors";
+    }
+
+    if(hand2 !== 'rock' && hand2 !== 'paper' && hand2 !== 'scissors') {
+      return "invalid entry for hand2 - must be either paper, rock, or scissors";
+    }
+
+    if(hand1 === hand2) {
```
Comment (tjefferson08) [inline apps/02RockPaperScissors.js:22]: Super simple. Nice!

File: apps/02RockPaperScissors_computer.js [hunk 17]
```diff
@@ -0,0 +1,121 @@
...
+var prompt = require('prompt');
+prompt.start();
+
+//first, start the game
+getPrompt();
+
+function getPrompt() {
+
+    //does the person want to play against someone else, or the computer?
+    prompt.get(['enter [1] for human vs human OR enter [2] for human vs computer'], function(error, result) {
```
Comment (tjefferson08) [inline apps/02RockPaperScissors_computer.js:13]: ooooo fun!

File: apps/03TicTacToe.js [hunk 18]
```diff
@@ -4,13 +4,23 @@ var assert = require('assert');
 var prompt = require('prompt');
 prompt.start();
 
+//general idea
```
Comment (tjefferson08) [inline apps/03TicTacToe.js:7]: Love this simple spec! Neat

File: apps/03TicTacToe.js [hunk 19]
```diff
@@ -23,30 +33,139 @@ function printBoard() {
...
+        return false;
+    }
+
+
 }
 
 function verticalWin() {
     // Your code here
+
+    if ((board[0][0] === 'X' && board[1][0] === 'X' && board[2][0] === 'X') || (board[0][1] === 'X' && board[1][1] === 'X' && board[2][1] === 'X') || (board[0][2] === 'X' && board[1][2] === 'X' && board[2][2] === 'X')) {
```
Comment (tjefferson08) [inline apps/03TicTacToe.js]: This if and the else if have _very_ similar logic. Maybe we could pull into a local function?!?!?!?

Ask me for deets if you want to go hardcore mode (you better)

File: apps/03TicTacToe.js [hunk 20]
```diff
@@ -23,30 +33,139 @@ function printBoard() {
...
+
+    horizontalWin();
+    verticalWin();
+    diagonalWin();
+
+    //if any are true, then there is a win!
+    //if not continue playing the game
+
+    if (horizontalWin() === true || verticalWin() === true || diagonalWin() === true) {
+        playerTurn = (playerTurn === '1') ? '2' : '1';
```
Comment (tjefferson08) [inline apps/03TicTacToe.js:120]: 1/2 instead of X/O interesting

File: apps/03TicTacToe.js [hunk 21]
```diff
@@ -56,28 +175,48 @@ function getPrompt() {
 
 if (typeof describe !== 'undefined') {
 
-    describe('#ticTacToe()', function () {
-        it('should place mark on the board', function () {
+    describe('#ticTacToe()', function() {
+        it('should place mark on the board', function() {
             ticTacToe(1, 1);
-            assert.deepEqual(board, [ [' ', ' ', ' '], [' ', 'X', ' '], [' ', ' ', ' '] ]);
+            assert.deepEqual(board, [
```
Comment (tjefferson08) [inline apps/03TicTacToe.js:212]: LOL cleanin' up the tests. Aight

File: apps/03TicTacToe.js (hunk 19 above)
Comment (nina2cool) [inline apps/03TicTacToe.js]: Sir, yes, Sir!

i will look at it

File: homework/04LessonFour.js [hunk 22]
```diff
@@ -87,7 +105,7 @@ var nestedObjects = {
 // What is an associative array? What is the difference between an array and an associative array?
 //
 // Your Answer Goes Here:
-//
+// An associative array (also called an object) is a collection of data - similar to a regular array.  However, for an associative array you can define the index explicitly.  A regular array the index is implicit.  This is handy in case you want to name the index instead of referring to the number.
```
Comment (tjefferson08) [inline homework/04LessonFour.js:108]: Great answer!

File: apps/04TowersOfHanoi.js [hunk 23]
```diff
@@ -10,74 +10,209 @@ var stacks = {
...
+    var lengthA = stacks.a.length;
+    var lengthB = stacks.b.length;
+    var lengthC = stacks['c'].length;
+
+    //adjust the array to move the last number from the start stack and add to the front of the end stack
+    var lastValueA = stacks.a[lengthA - 1];
+    var lastValueB = stacks.b[lengthB - 1];
+    var lastValueC = stacks.c[lengthC - 1];
+
+    if (startStack === 'a') {
```
Comment (tjefferson08) [inline apps/04TowersOfHanoi.js]: This looks clean, but a little repetitive. What if we had 5 or 7 or 100 stacks? Would we write 100 `else if` blocks?

File: apps/04TowersOfHanoi.js [hunk 24]
```diff
@@ -10,74 +10,209 @@ var stacks = {
...
+        lastValueStartStack = 0;
+    }
+    if (lengthEndStack === 0) {
+        lastValueEndStack = 0;
+    }
+
+    // move is legal if the value you are moving to the endStack is LESS than the last value of the startStack
+    console.log('lastValueStartStack = ' + lastValueStartStack + ' and lastValueEndStack = ' + lastValueEndStack)
+
+    if (lastValueEndStack === 0) {
```
Comment (tjefferson08) [inline apps/04TowersOfHanoi.js:56]: I think I can come up with an illegal move here. What if start and end stack are _both_ empty?

File: apps/04TowersOfHanoi.js [hunk 25]
```diff
@@ -10,74 +10,209 @@ var stacks = {
...
+        return true;
+    } else if (lastValueStartStack > lastValueEndStack) {
+        return false;
+    } else if (lastValueStartStack < lastValueEndStack) {
```
Comment (tjefferson08) [inline apps/04TowersOfHanoi.js:60]: Here we check > and <, what happens if they're === ?

File: apps/04TowersOfHanoi.js [hunk 26]
```diff
@@ -10,74 +10,209 @@ var stacks = {
...
+    // FIRST check isLegal
+    // If isLegal is true, then do movePiece.  If not true, ask for prompt again.
+    // THIRD checkForWin - if win, end game.  if no win, continue function.
+    // FOURTH if no winner, increase numberOfTurns & continue game
+
+    var isLegalFunction = isLegal(startStack, endStack);
+
+    console.log('Is the move legal? ' + isLegalFunction);
+
+    if (isLegalFunction === true) {
```
Comment (tjefferson08) [inline apps/04TowersOfHanoi.js]: Checking === true isn't usually necessary. What datatype is the variable `isLegalFunction` ?

File: apps/04TowersOfHanoi.js [hunk 27]
```diff
@@ -10,74 +10,209 @@ var stacks = {
...
-        towersOfHanoi(result['start stack'], result['end stack']);
-        getPrompt();
+    printNumberOfTurns(numberOfTurns);
+    prompt.get(['start stack', 'end stack'], function(error, result) {
+
+        var startStack = result['start stack'];
+        var endStack = result['end stack'];
+        var lengthStartStack = stacks[startStack].length;
+
+        if ((startStack != 'a' && startStack != 'b' && startStack != 'c') || (endStack != 'a' && endStack != 'b' && endStack != 'c')) {
```
Comment (tjefferson08) [inline apps/04TowersOfHanoi.js:122]: Cool extensions!

File: apps/04TowersOfHanoi.js (hunk 24 above)
Comment (nina2cool) [inline apps/04TowersOfHanoi.js:56]: On line 128 in the GetPrompt I don't let the user put in the same value for the start stack and the end stack... so when this function runs, it has already prevented it.  Can you check that?

File: apps/04TowersOfHanoi.js (hunk 25 above)
Comment (nina2cool) [inline apps/04TowersOfHanoi.js:60]: I check in the beginning in GetPrompt if they are equal.  If they aren't, the user must enter a new value.  Also, GetPrompt also checks if the stack chosen is empty - the user cannot choose two empty stacks.  Therefore, if the starting stack has zero duplicate values, then this case doesn't come up.

File: homework/05LessonFive.js [hunk 28]
```diff
@@ -63,15 +63,15 @@ for (forLoopCounter = 0; forLoopCounter > -10; forLoopCounter--) {
...
+var forLoopArrayResult = [0, -1, -2, -3, -4, -5, -6, -7, -8, -9];
 
 // ****
 // Concept Checkpoint
 //
 // Write your answer in comments
 //
 // What are loops? Why do we use them?
-//
+// ANSWER: A loop is code written to perform a repetitive task when you know how many times you want to repeat the task.  We use them so we don't have to write lengthy code or do repetive tasks.
```
Comment (tjefferson08) [inline homework/05LessonFive.js:74]: Nice! Do we ever _not_ know how  many times it'll run?
What if I use `Math.rand()` or get input from `prompt`?

File: apps/05Mastermind.js [hunk 29]
```diff
@@ -6,68 +6,169 @@ var prompt = require('prompt');
...
+        var letterIndex = solutionArray.indexOf(guessArray[j]);
+
+        if (letterIndex > -1) {
+            correctLetters++;
+        }
+
+    }
+
+    //we don't want to double count the correct letter positions, so we subtract that number out
+    correctLetters = correctLetters - correctLetterLocations;
```
Comment (tjefferson08) [inline apps/05Mastermind.js:75]: Interesting! Another way to exclude double-counts would be to add some more logic above on line 68. Can you think of what that might be? Why are we getting double-counts even though we're nulling out the guessArray elements?

File: apps/05Mastermind.js [hunk 30]
```diff
@@ -6,68 +6,169 @@ var prompt = require('prompt');
...
+    //advance the number of guesses by 1
+    numberOfGuesses++;
+
+    printNumberOfGuesses();
+
+    if (numberOfGuesses < 10) {
+
+        if (guess === solution) {
+
+            //I had fancier code here to customize the response if they guessed in 1 try - but returning anything other than "You guessed it!" means it wouldn't pass the test.  So I am not using this code right now.
```
Comment (tjefferson08) [inline apps/05Mastermind.js:101]: Nice reminder

File: apps/05Mastermind.js [hunk 31]
```diff
@@ -6,68 +6,169 @@ var prompt = require('prompt');
...
-        console.log( mastermind(result['guess']) );
-        printBoard();
-        getPrompt();
+
+    prompt.get(['guess'], function(error, result) {
+
+        var guess = result['guess'];
+        var guessLength = guess.length;
+
+        if (guessLength === 4) {
```
Comment (tjefferson08) [inline apps/05Mastermind.js:133]: Sweet validations

File: homework/06LessonSix.js [hunk 32]
```diff
@@ -89,14 +97,15 @@ var conferenceRoom = new ConferenceRoom();
...
 // add your code for Problem 6 here
 
+conferenceRoom.enter(jen);
 
 // ****
 // Concept Checkpoint
 //
 // Write your answer in comments
 //
 // What is a method?
```
[the remark this answers is not in the document]
Comment (tjefferson08) [reply homework/06LessonSix.js:107]: lol ^

File: homework/06LessonSix.js [hunk 33]
```diff
@@ -89,14 +97,15 @@ var conferenceRoom = new ConferenceRoom();
...
-//
+//  An action that can be performed on an object.  Defined as a function within the object class.
```
Comment (tjefferson08) [inline homework/06LessonSix.js:108]: yep yep!

File: apps/06SpaceTravelToMars.js [hunk 34]
```diff
@@ -3,57 +3,138 @@
...
+            if (countValidJobs > 0) {
+                return this.ability;
+            } else {
+                return "Can't perform a mission yet.";
+            }
+        }
+    }
+}
+
+var mav = new Ship('Mars Ascent Vehicle', 'MAV', 'Ascend into low orbit');
```
Comment (tjefferson08) [inline apps/06SpaceTravelToMars.js:73]: Don't need this but it's not hurting anything

File: apps/06SpaceTravelToMars.js [hunk 35]
```diff
@@ -3,57 +3,138 @@
...
+
+            for (var i = 0; i < this.crew.length; i++) {
+
+                if (jobTypes[this.crew[i]['job']] == this.type) {
+                    countValidJobs++;
+                }
+            }
+
+            //Look at countValidJobs variable to see if there are any valid jobs. (should be greater than zero)
+            if (countValidJobs > 0) {
```
Comment (tjefferson08) [inline apps/06SpaceTravelToMars.js:64]: I like this approach. If you can remember, find me in class and let's talk about if block nesting and 'early exit' scenarios

File: homework/07LessonSeven/script.js [hunk 36]
```diff
@@ -2,87 +2,111 @@
...
+    // ****
+
+    // Problem 9:
+    // select the element with id="insert-text"
+    // then set the text of this element to "Text Has Been Inserted"
+    window.insertText = $('#insert-text');
+    window.insertText.text('Text Has Been Inserted');
+
+    //better way
+    $('#insert-text').text('Text Has Been Inserted');
```
[the remark this answers is not in the document]
Comment (tjefferson08) [reply homework/07LessonSeven/script.js:78]: Better indeed! Although it's occasionally good to remember that all $() calls 'cost' something, so storing them in variables can improve performance

File: apps/07jquery-tic-tac-toe/script.js [hunk 37]
```diff
@@ -2,4 +2,177 @@
...
+                var cellContents1 = $("[data-cell='1']").text();
+                var cellContents2 = $("[data-cell='2']").text();
+                var cellContents3 = $("[data-cell='3']").text();
+                var cellContents4 = $("[data-cell='4']").text();
+                var cellContents5 = $("[data-cell='5']").text();
+                var cellContents6 = $("[data-cell='6']").text();
+                var cellContents7 = $("[data-cell='7']").text();
+                var cellContents8 = $("[data-cell='8']").text();
+
+                checkForWin(cellContents0, cellContents1, cellContents2, cellContents3, cellContents4, cellContents5, cellContents6, cellContents7, cellContents8);
```
Comment (tjefferson08) [inline apps/07jquery-tic-tac-toe/script.js]: We talked about this in class, I know you're already hard at work improving it :D

File: apps/07jquery-tic-tac-toe/script.js [hunk 38]
```diff
@@ -2,4 +2,177 @@
...
+        }
+
+    }
+
+    //clear the board
+    var button = $('#clear');
+
+    $('#clear').click(function() {
+
+        $("[data-cell='0']").text('');
```
Comment (tjefferson08) [inline apps/07jquery-tic-tac-toe/script.js:210]: Is there any way to select all these at once? Or maybe loop?

**Pull request closed without merging**